doc/xxd.1
.TH XXD 1 "August 1996" "Manual page for xxd"
.\"
.\" 21st May 1996
.\" Man page author:
.\"    Tony Nugent <tony@sctnugen.ppp.gu.edu.au> <T.Nugent@sct.gu.edu.au>
.\"    Changes by Bram Moolenaar <mool@oce.nl>
.SH NAME
.I xxd
\- make a hexdump or do the reverse.
.SH SYNOPSIS
.B xxd \-h
[
.B elp
]
.br
.B xxd
[
.I options
] [
.I infile
[
.I outfile
]]
.br
.B xxd \-r
[
.B evert
] [
.I options
] [
.I infile
[
.I outfile
]]
.SH DESCRIPTION
.I xxd
creates a hex dump of a given file or standard input.
It can also convert a hex dump back to its original binary form.
Like
.BR uuencode(1)
and
.BR uudecode(1)
it allows the transmission of binary data in a `mail-safe' ASCII representation,
but has the advantage of decoding to standard output.
Moreover, it can be used to perform binary file patching.
.SH OPTIONS
If no
.I infile
is given, standard input is read.
If
.I infile
is specified as a
.RB \` \- \'
character, then input is taken from standard input.
If no
.I outfile
is given (or a 
.RB \` \- \'
character is in its place), results are sent to standard output.
.PP
Note that a "lazy" parser is used which does not check for more than the first
letter, unless the option is followed by a parameter.
Spaces between an option and a following parameter is optional,
and in that case the long name is known to make equivalent
\fB\-c8\fR, \fB\-c 8\fR, \fB\-cols 8\fR, and \fB\-cols8\fR.
.PP
.TP
.I \-a \-autoskip
toggle autoskip: A single '*' replaces nul-lines.  Default off.
(This feature may not work correctly on some platforms).
.TP
\fI\-c cols \fR| \fI\-cols cols
format
.RI < cols >
octets per line. Default 16 (-i: 12, -ps: 30). Max 64.
.TP
.I \-h \-help
print a summary of available commands and exit.  No hex dumping is performed.
.TP
.I \-i \-include
output in C include file style.
.TP
\fI\-l len \fR| \fI\-len len
stop after
.RI  < len >
octets.
.TP
.I \-p \-ps \-postscript
output in postscript continuous hexdump style.
.TP
.I \-r \-revert
reverse operation: convert (or patch) hexdump into binary.
.TP
.I \-seek off
When used after
.I \-r
: revert with
.RI < offset >
added to file positions found in hexdump.
.TP
.I \-s [\+][\-]seek
start at <seek> bytes abs. (or +: rel.) infile offset.
.I seek
may be hex (eg, 0x3f) or decimal (eg, 47).
\fI\+ \fRindicates that the seek is from the start of the input.
\fI\- \fRindicates that the seek should be that many characters from the end of the input.
.TP
.I \-u
use upper case hex letters (default is lower case).
.TP
.I \-v
show version string.
.SH CAVEATS
.PP
.I xxd \-r
has some builtin magic.
The linenumbers at the start of each line matter!
.PP
Note that there is a difference between
.br
% xxd \-i file
.br
and
.br
% xxd \-i \< file
.PP
.I xxd \-s \+seek
has some semantic differences, as lseek() is used to "rewind" input.  A '+'
makes a difference if the input source is stdin, and if stdin's file position
is not at the start of the file by the time xxd is started and given its input.
The following examples may help to clarify (or further confuse!)...
.PP
.I Rewind stdin before reading; needed because the `cat' has already read to the end of stdin.
.br
% sh -c 'cat > plain_copy; xxd -s 0 > hex_copy' < file
.PP
.I Hexdump from file position 0x480 (=1024+128) onwards.
.I The `+' sign means "relative to the current position", thus the `128' adds to the 1k where dd left off.
.br
% sh -c 'dd of=plain_snippet bs=1k count=1; xxd -s +128 > hex_snippet' < file
.PP
.I Hexdump from file position 0x100 ( = 1024-768) on. 
.br
% sh -c 'dd of=plain_snippet bs=1k count=1; xxd -s +-768 > hex_snippet' < file
.PP
However, this is a rare situation and the use of `+' is rarely needed.
.SH EXAMPLES
.PP
.br
.I Print everything but the first three lines (hex 0x30 bytes) of
.B file
\.
.br
% xxd \-s 0x30 file
.PP
.br
.I Print 3 lines (hex 0x30 bytes) from the end of
.B file
\.
.br
% xxd \-s \-0x30 file
.PP
.br
.I Print 120 bytes as continuous hexdump with 40 octets per line.
.br
% xxd \-l 120 \-ps \-c 20 xxd.1
.br
2e544820585844203120224d616e75616c207061
.br
676520666f7220787864220a2e5c220a2e5c2220
.br
32317374204d617920313939360a2e5c22204d61
.br
6e207061676520617574686f723a0a2e5c222020
.br
2020546f6e79204e7567656e74203c746f6e7940
.br
7363746e7567656e2e7070702e67752e6564752e
.br

.br
.I Hexdump the first 120 bytes of this man page with 12 octets per line.
.br
% xxd \-l 120 \-c 12 xxd.1
.br
0000000: 2e54 4820 5858 4420 3120 224d  .TH XXD 1 "M
.br
000000c: 616e 7561 6c20 7061 6765 2066  anual page f
.br
0000018: 6f72 2078 7864 220a 2e5c 220a  or xxd"..\\".
.br
0000024: 2e5c 2220 3231 7374 204d 6179  .\\" 21st May
.br
0000030: 2031 3939 360a 2e5c 2220 4d61   1996..\\" Ma
.br
000003c: 6e20 7061 6765 2061 7574 686f  n page autho
.br
0000048: 723a 0a2e 5c22 2020 2020 546f  r:..\\"    To
.br
0000054: 6e79 204e 7567 656e 7420 3c74  ny Nugent <t
.br
0000060: 6f6e 7940 7363 746e 7567 656e  ony@sctnugen
.br
000006c: 2e70 7070 2e67 752e 6564 752e  .ppp.gu.edu.
.PP
.br
.I Display just the date from the file xxd.1
.br
% xxd \-s 0x28 \-l 12 \-c 12 xxd.1
.br
0000028: 3231 7374 204d 6179 2031 3939  21st May 199
.PP
.br
.I Copy
.B input_file
.I to
.B output_file
.I and prepend 100 bytes of value 0x00.
.br
% xxd input_file | xxd \-r \-s 100 \> output_file
.br

.br
.I Patch the date in the file xxd.1
.br
% echo '0000029: 3574 68' | xxd -r - xxd.1
.br
% xxd \-s 0x28 \-l 12 \-c 12 xxd.1
.br
0000028: 3235 7468 204d 6179 2031 3939  25th May 199
.PP
.br
.I Create a 65537 byte file with all bytes 0x00,
.I except for the last one which is 'A' (hex 0x41).
.br
% echo \'010000: 41\' | xxd \-r \> file
.PP
.br
.I Hexdump this file with autoskip.
.br
% xxd \-a \-c 12 file
.br
0000000: 0000 0000 0000 0000 0000 0000  ............
.br
*
.br
000fffc: 0000 0000 40                   ....A
.br

.br
.I Create a 1 byte file containing a single 'A' character.
.I The number after '-r -s' adds to the linenumbers found in the file;
.I in effect, the leading bytes are suppressed.
.br
% echo '010000: 41' | xxd \-r \-s \-0x10000 \> file
.PP
.br
.I Use xxd as a filter within an editor such as
.B vim(1)
.I to hexdump a region marked between `a' and `z'.
.br
:'a,'z!xxd
.br

.br
.I Use xxd as a filter within an editor such as
.B vim(1)
.I to recover a binary hexdump marked between `a' and `z'.
.br
:'a,'z!xxd -r
.br

.br
.I Use xxd as a filter within an editor such as
.B vim(1)
.I to recover one line of a hexdump.  Move the cursor over the line and type:
.br
!!xxd -r
.br
.SH "RETURN VALUES"
The following error values are returned:
.TP
0
no errors encountered.
.TP
1
parse error.
.SH "SEE ALSO"
uuencode(1), uudecode(1), patch(1)
.br
.SH WARNINGS
Use entirely at your own risk.
.br
.SH BUGS
.br
\fB\-a\fR may not work properly on all platforms.
.br
\fB\-version\fR does not work correctly (use \fB\-v\fR).
.br
\fB\-seek3\fR and \fB\-skip3\fR (and others) do not work
(use a space between the option and its parameter).
.br
.SH AUTHOR
.br
(c) 1990-1996 by Juergen Weigert
.br
<jnweiger@informatik.uni-erlangen.de>
.LP
Distribute freely and credit me,
.br
make money and share with me,
.br
lose money and don't ask me.
.PP
Manual page by Tony Nugent
.br
<tony@sctnugen.ppp.gu.edu.au> <T.Nugent@sct.gu.edu.au>
.PP


src/xxd/xxd.c
/* xxd: my hexdump facility. jw
 * changed to word output 2.10.90
 * new indent style, dumb bug inserted and fixed. 3.3.93
 * -c option, mls
 * better option parser, -ps, -l, -s  added, 26.4.94
 * -r badly needs - as input file, added. 1.7.94
 * Per default autoskip over consequtive lines of zeroes, as unix od does.
 * -a shows them too.
 * -i dump as c-style #include "file.h"
 *    if "xxd -i" knows the filename, an 'unsigned char filename_bits[]' array
 *    is written in correct c-syntax. added 1.11.95
 * -s improved, now defaults to absolute seek, relative requires a '+'.
 * -r improved, now -r -s -0x... is supported change/suppress leading \0 bytes.
 * -l n improved: stops exactly after n bytes, even if partial lines result.
 * -r improved, better handling of partial lines with trailing garbage.
 * -r improved, now -r -p works again!
 * -r improved, less flushing, much faster now! (that was silly)
 * Per repeated request of a single person: autoskip defaults to off.
 *	3.4.96. jw.
 * -v added. They want to know the version. 15.5.96
 * -a fixed, to show last line inf file ends in all zeros.
 * -u added: Print upper case hex-letters, as preferred by unix bc.
 * -h added to usage message. Usage message extended.
 * Now using outfile if specified even in normal mode, aehem.
 * No longer mixing of ints and longs. May help doze people.
 * Added binify ioctl for same reason. (Enough Doze stress for 1996!)
 * -p improved, removed occasional superfluous linefeed.  16.5.96
 * -l 0 fixed. tried to read anyway. 20.5.96
 * -i fixed. now honours -u, and prepends __ to numeric filenames. 21.5.96
 * compile -DWIN32 for NT or W95. George V. Reilly, * -v improved :-)
 * support --gnuish-longhorn-options
 * MAC support added: CodeWarrior already uses ``outline'' in Types.h which is
 *     included by MacHeaders (Axel Kielhorn 1996/05/25). Renamed to xxdline().
 *     jw.
 * -i printed 'int' instead of 'char'. *blush* 7.6.96
 * added bram's OS2 ifdefs...
 *
 * (c) 1990-1996 by Juergen Weigert (jnweiger@informatik.uni-erlangen.de)
 *
 * Distribute freely and credit me,
 * make money and share with me,
 * lose money and don't ask me.
 */
#include <stdio.h>
#include <fcntl.h>
#ifdef __TSC__
# define MSDOS
#endif
#if !defined(OS2) && defined(__EMX__)
# define OS2
#endif
#if defined(MSDOS) || defined(WIN32) || defined(OS2)
# include <io.h>	/* for setmode() */
#elif defined(UNIX)
# include <unistd.h>
#endif
#include <stdlib.h>
#include <string.h>	/* for strncmp() */
#include <ctype.h>	/* for isalnum() */
#if __MWERKS__
# include <unix.h>	/* for fdopen() on MAC */
#endif

extern long int strtol();
extern long int ftell();

#ifdef WIN32
char version[] = "xxd V1.4m 8aug96 (Win32) by Juergen Weigert";
#elif defined DJGPP
char version[] = "xxd V1.4m 8aug96 (32 bit) by Juergen Weigert";
#elif defined MSDOS
char version[] = "xxd V1.4m 8aug96 (16 bit) by Juergen Weigert";
#else
char version[] = "xxd V1.4m 8aug96 by Juergen Weigert";
#endif

#if defined(MSDOS) || defined(WIN32) || defined(OS2)
# define BIN_READ(yes)  ((yes) ? "rb" : "rt")
# define BIN_WRITE(yes) ((yes) ? "wb" : "wt")
# define BIN_ASSIGN(fp, yes) setmode(fileno(fp), (yes) ? O_BINARY : O_TEXT)
#else
# define BIN_READ(dummy)  "r"
# define BIN_WRITE(dummy) "w"
# define BIN_ASSIGN(fp, dummy) fp
#endif

/* open has only to arguments on the Mac */
#if __MWERKS__		
# define OPEN(name, mode, umask) open(name, mode)
#else
# define OPEN(name, mode, umask) open(name, mode, umask)
#endif

#ifndef __P
# ifdef __STDC__
#  define __P(a) a
# else
#  define __P(a) ()
# endif
#endif

/* Let's collect some prototypes */
/* CodeWarrior is really picky about missing prototypes */
static void exit_with_usage __P((char *));
static int huntype __P((FILE *, FILE *, FILE *, char *, int, int, long));
static void xxdline __P((FILE *, char *, int));

#define TRY_SEEK 	/* attempt to use lseek, or skip forward by reading */
#define COLS 64		/* change here, if you ever need more columns */
#define LLEN (9 + (5*COLS-1)/2 + 2 + COLS)

char hexxa[] = "0123456789abcdef0123456789ABCDEF", *hexx = hexxa;

/* the different hextypes known by this program: */
#define HEX_NORMAL 0
#define HEX_POSTSCRIPT 1
#define HEX_CINCLUDE 2

static void
exit_with_usage(pname)
char *pname;
{
  fprintf(stderr, "Usage:\n       %s [options] [infile [outfile]]\n", pname);
  fprintf(stderr, "    or\n       %s -r [-s [-]offset] [-c cols] [-ps] [infile [outfile]]\n", pname);
  fprintf(stderr, "Options:\n");
  fprintf(stderr, "    -a          toggle autoskip: A single '*' replaces nul-lines. Default off.\n");
  fprintf(stderr, "    -c cols     format <cols> octets per line. Default 16 (-i: 12, -ps: 30).\n");
  fprintf(stderr, "    -h          print this summary.\n");
  fprintf(stderr, "    -i          output in C include file style.\n");
  fprintf(stderr, "    -l len      stop after <len> octets.\n");
  fprintf(stderr, "    -ps         output in postscript continuous hexdump style.\n");
  fprintf(stderr, "    -r          reverse operation: convert (or patch) hexdump into binary.\n");
  fprintf(stderr, "    -r -s off   revert with <off> added to file positions found in hexdump.\n");
  fprintf(stderr, "    -s %sseek  start at <seek> bytes abs. %sinfile offset.\n", 
#ifdef TRY_SEEK
	  "[+][-]", "(or +: rel.) ");
#else
	  "", "");
#endif
  fprintf(stderr, "    -u          use upper case hex letters.\n");
  fprintf(stderr, "    -v          show version: \"%s\".\n", version);
  exit(1);
}

/* 
 * Max. cols binary characters are decoded from the input stream per line.
 * Two adjacent garbage characters after evaluated data delimit valid data.
 * Everything up to the next newline is discarded. 
 *
 * The name is historic and came from 'undo type opt h'.
 */
static int
huntype(fpi, fpo, fperr, pname, cols, hextype, base_off)
FILE *fpi, *fpo, *fperr;
char *pname;
int cols, hextype;
long base_off;
{
  int c, ign_garb = 1, n1 = -1, n2 = 0, n3, p = cols;
  long have_off = 0, want_off = 0;

  rewind(fpi);

  while ((c = getc(fpi)) != EOF)
    {
      if (c == '\r')	/* Doze style input file? */
        continue;

      n3 = n2;
      n2 = n1;

      if (c >= '0' && c <= '9')
        n1 = c - '0';
      else if (c >= 'a' && c <= 'f')
        n1 = c - 'a' + 10;
      else if (c >= 'A' && c <= 'F')
        n1 = c - 'A' + 10;
      else 
        {
          n1 = -1;
	  if (ign_garb)
	    continue;
	}

      ign_garb = 0;

      if (p >= cols)
	{
	  if (!hextype)
	    {
	      if (n1 < 0)
		{
		  p = 0;
		  continue;
		}
	      want_off = (want_off << 4) | n1;
	      continue;
	    }
	  else
	    p = 0;
	}

      if (base_off + want_off != have_off)
        {
	  fflush(fpo);
#ifdef TRY_SEEK
	  c = fseek(fpo, base_off + want_off - have_off, 1);
	  if (c >= 0)
	    have_off = base_off + want_off;
#endif
	  if (base_off + want_off < have_off)
	    {
	      fprintf(fperr, "%s: sorry, cannot seek backwards.\n", pname);
	      return 5;
	    }
	  for (; have_off < base_off + want_off; have_off++)
	    putc(0, fpo);
	}

      if (n2 >= 0 && n1 >= 0)
        {
	  putc((n2 << 4) | n1, fpo);
	  have_off++;
	  want_off++;
	  n1 = -1;
	  if (++p >= cols)
	    {
	      if (!hextype)
	        want_off = 0;
	      while ((c = getc(fpi)) != '\n' && c != EOF)
	        ;
	      ign_garb = 1;
	    }
	}
      else if (n1 < 0 && n2 < 0 && n3 < 0)
        {
	  if (!hextype)
	    want_off = 0;
	  while ((c = getc(fpi)) != '\n' && c != EOF)
	    ;
	  ign_garb = 1;
	}
    }
  fflush(fpo);
#ifdef TRY_SEEK
  fseek(fpo, 0L, 2);
#endif
  fclose(fpo);
  fclose(fpi);
  return 0;
}

/*
 * Print line l. If nz is false, xxdline regards the line a line of 
 * zeroes. If there are three or more consecutive lines of zeroes,
 * they are replaced by a single '*' character. 
 *
 * If the output ends with more than two lines of zeroes, you
 * should call xxdline again with l being the last line and nz 
 * negative. This ensures that the last line is shown even when
 * it is all zeroes.
 * 
 * If nz is always positive, lines are never suppressed.
 */
static void
xxdline(fp, l, nz)
FILE *fp;
char *l;
int nz;
{
  static char z[LLEN+1];
  static int zero_seen = 0;

  if (!nz && zero_seen == 1)
    strcpy(z, l);

  if (nz || !zero_seen++)
    {
      if (nz)
	{
	  if (nz < 0)
	    zero_seen--;
	  if (zero_seen == 2)
	    fputs(z, fp);
	  if (zero_seen > 2)
	    fputs("*\n", fp);
	}
      if (nz >= 0 || zero_seen > 0)
        fputs(l, fp);
      if (nz)
	zero_seen = 0;
    }
}

int
main(argc, argv)
int argc;
char *argv[];
{
  FILE *fp, *fpo;
  int c, e, p = 0, relseek = 1, negseek = 0, hextype, revert = 0;
  int autoskip = 0;
  long length = -1, n = 0, seekoff = 0;
  char l[LLEN+1];
  int cols = 0, nonzero = 0;
  char *pname, *pp;
  hextype = HEX_NORMAL;

  pname = argv[0];
  for (pp = pname; *pp; )
    if (*pp++ == '/')
      pname = pp;

  while (argc >= 2)
    {
      pp = argv[1] + (!strncmp(argv[1], "--", (size_t)2) && argv[1][2]);
           if (!strncmp(pp, "-a", (size_t)2)) autoskip = 1 - autoskip;
      else if (!strncmp(pp, "-u", (size_t)2)) hexx = hexxa + 16;
      else if (!strncmp(pp, "-p", (size_t)2)) hextype = HEX_POSTSCRIPT;
      else if (!strncmp(pp, "-i", (size_t)2)) hextype = HEX_CINCLUDE;
      else if (!strncmp(pp, "-r", (size_t)2)) revert++;
      else if (!strncmp(pp, "-v", (size_t)2)) 
        {
	  fprintf(stderr, "%s\n", version);
	  exit(0);
	}
      else if (!strncmp(pp, "-c", (size_t)2))
	{
	  if (pp[2] && strncmp("ols", pp + 2, (size_t)3))
	    cols = (int)strtol(pp + 2, NULL, 0);
	  else
	    {
	      if (!argv[2])
	        exit_with_usage(pname);
	      cols = (int)strtol(argv[2], NULL, 0);
	      argv++;
	      argc--;
	    }
	}
      else if (!strncmp(pp, "-s", (size_t)2))
	{
	  relseek = 0;
	  negseek = 0;
	  if (pp[2] && strncmp("kip", pp+2, (size_t)3) && strncmp("eek", pp+2, (size_t)3))
	    {
#ifdef TRY_SEEK
	      if (pp[2] == '+')
	        relseek++;
	      if (pp[2+relseek] == '-')
	        negseek++;
#endif
	      seekoff = strtol(pp + 2+relseek+negseek, (char **)NULL, 0);
	    }
	  else
	    {
	      if (!argv[2])
	        exit_with_usage(pname);
#ifdef TRY_SEEK
	      if (argv[2][0] == '+')
	        relseek++;
	      if (argv[2][relseek] == '-')
	        negseek++;
#endif
	      seekoff = strtol(argv[2] + relseek+negseek, (char **)NULL, 0);
	      argv++;
	      argc--;
	    }
	}
      else if (!strncmp(pp, "-l", (size_t)2))
	{
	  if (pp[2] && strncmp("en", pp + 2, (size_t)2))
	    length = strtol(pp + 2, (char **)NULL, 0);
	  else
	    {
	      if (!argv[2])
	        exit_with_usage(pname);
	      length = strtol(argv[2], (char **)NULL, 0);
	      argv++;
	      argc--;
	    }
	}
      else if (!strcmp(pp, "--"))	/* end of options */
        {
	  argv++;
	  argc--;
	  break;
	}
      else if (pp[0] == '-' && pp[1])	/* unknown option */
        exit_with_usage(pname);
      else 
        break;				/* not an option */

      argv++;				/* advance to next argument */
      argc--;
    }

  if (!cols)
    switch (hextype)
      {
      case HEX_POSTSCRIPT: 	cols = 30; break;
      case HEX_CINCLUDE:	cols = 12; break;
      case HEX_NORMAL:
      default:			cols = 16; break;
      }

  if (cols < 1 || (!hextype && (cols > COLS)))
    {
      fprintf(stderr, "%s: invalid number of columns (max. %d).\n", pname, COLS);
      exit(1);
    }

  if (argc > 3) 
    exit_with_usage(pname);

  if (argc == 1 || (argv[1][0] == '-' && !argv[1][1]))
    BIN_ASSIGN(fp = stdin, !revert);
  else
    {
      if ((fp = fopen(argv[1], BIN_READ(!revert))) == NULL)
	{
	  fprintf(stderr,"%s: ", pname);
	  perror(argv[1]);
	  return 2;
	}
    }
  
  if (argc < 3 || (argv[2][0] == '-' && !argv[2][1]))
    BIN_ASSIGN(fpo = stdout, revert);
  else
    {
      int fd;

      if (((fd = OPEN(argv[2], O_WRONLY | O_CREAT, 0666)) < 0) ||
          (fpo = fdopen(fd, BIN_WRITE(revert))) == NULL)
	{
	  fprintf(stderr, "%s: ", pname);
	  perror(argv[2]);
	  return 3;
	}
      rewind(fpo);
    }
  
  if (revert)
    {
      if (hextype && (hextype != HEX_POSTSCRIPT))
        {
          fprintf(stderr, "%s: sorry, cannot revert this type of hexdump\n", pname);
	  return -1;
	}
      return huntype(fp, fpo, stderr, pname, cols, hextype, 
      		negseek ? -seekoff : seekoff);
    }

  if (seekoff || negseek || !relseek)
    {
#ifdef TRY_SEEK
      if (relseek)
        e = fseek(fp, negseek ? -seekoff : seekoff, 1);
      else
        e = fseek(fp, negseek ? -seekoff : seekoff, negseek ? 2 : 0);
      if (e < 0 && negseek)
        {
	  fprintf(stderr, "%s: sorry cannot seek.\n", pname);
	  return 4;
	}
      if (e >= 0)
        seekoff = ftell(fp);
      else
#endif
	{
	  long s = seekoff;

	  while (s--)
	    getc(fp); 
	}
    }

  if (hextype == HEX_CINCLUDE)
    {
      if (fp != stdin)
	{
	  fprintf(fpo, "unsigned char %s", isdigit(argv[1][0]) ? "__" : "");
	  for (e = 0; (c = argv[1][e]); e++)
	    putc(isalnum(c) ? c : '_', fpo);
	  fputs("[] = {\n", fpo);
	}

      p = 0;
      while ((length < 0 || p < length) && (c = getc(fp)) != EOF)
	{
	  fprintf(fpo, (hexx == hexxa) ? "%s0x%02x" : "%s0X%02X",
	    (p % cols) ? ", " : ",\n  "+2*!p,  c);
	  p++;
	}

      if (p)
        fputs("\n};\n"+3*(fp == stdin), fpo);

      if (fp != stdin)
	{
	  fprintf(fpo, "unsigned int %s", isdigit(argv[1][0]) ? "__" : "");
	  for (e = 0; (c = argv[1][e]); e++)
	    putc(isalnum(c) ? c : '_', fpo);
	  fprintf(fpo, "_len = %d;\n", p);
	}

      fclose(fp);
      fclose(fpo);
      return 0;
    }

  if (hextype == HEX_POSTSCRIPT)
    {
      p = cols;
      while ((length < 0 || n < length) && (e = getc(fp)) != EOF)
	{
          putchar(hexx[(e >> 4) & 0xf]);
          putchar(hexx[(e     ) & 0xf]);
	  n++;
	  if (!--p)
	    {
	      putchar('\n');
	      p = cols;
	    }
	}
      if (p < cols)
	putchar('\n');
      fclose(fp);
      fclose(fpo);
      return 0;
    }

  /* hextype == HEX_NORMAL */

  while ((length < 0 || n < length) && (e = getc(fp)) != EOF)
    {
      if (p == 0)
	{
	  sprintf(l, "%07lx: ", n + seekoff);
	  for (c = 9; c < LLEN; l[c++] = ' ');
	}
      l[ 9 + (5 * p) / 2] = hexx[(e >> 4) & 0xf];
      l[10 + (5 * p) / 2] = hexx[ e       & 0xf];
      l[11 + (5 * cols - 1) / 2 + p] = (e > 31 && e < 127) ? e : '.';
      if (e) 
        nonzero++;
      n++;
      if (++p == cols)
	{
	  l[c = (11 + (5 * cols - 1) / 2 + p)] = '\n'; l[++c] = '\0';
	  xxdline(fpo, l, autoskip ? nonzero : 1);
	  nonzero = 0;
	  p = 0;
	}
    }
  if (p)
    {
      l[c = (11 + (5 * cols - 1) / 2 + p)] = '\n'; l[++c] = '\0';
      xxdline(fpo, l, 1);
    }
  else if (autoskip)
    xxdline(fpo, l, -1);	/* last chance to flush out supressed lines */
    
  fclose(fp);
  fclose(fpo);
  return 0;
}
